### PR TITLE
perf: fix quadratic manifest scan in quarantine rebuild

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary
- Port the quarantine rebuild-from-scratch path onto `scripts/retro_quarantine_self_pollution.py` while preserving incremental parity and unquarantine restoration.
- Replace hot `chunk_id` filters on FTS5 tables in `_record_manifest` and `capture_restore_state` with indexed `chunk_fts_rowids` rowid-map lookups.
- Make `batch_size` drive real manifest/reclassify batches with per-batch commits/checkpoints, followed by one rowid-preserving FTS rebuild.

## Dispatch Brief
- Local build brief: `/Users/etanheyman/Gits/brainlayer/docs.local/.build-briefs/2026-07-08-pr3-quadratic-fix.md`
- The brief is in `docs.local` outside this worktree, so the relevant requirement/evidence is included inline here rather than relying on a gitignored PR artifact.

## Query-Plan Evidence
- Added `tests/test_retro_self_pollution_quarantine.py::test_record_manifest_uses_indexed_fts_rowid_lookup_plan`.
- Added `tests/test_retro_self_pollution_quarantine.py::test_capture_restore_state_uses_indexed_fts_rowid_lookup_plan`.
- Both tests execute `EXPLAIN QUERY PLAN` against the SQL emitted by the script and assert the plan uses `chunk_fts_rowids` with `SEARCH`, while rejecting `SCAN chunks_fts VIRTUAL TABLE`, `SCAN chunks_fts_trigram VIRTUAL TABLE`, and `SCAN chunks_fts_operational VIRTUAL TABLE`.

## Test Plan
- [x] RED verified before implementation: the new query-plan tests failed on `SCAN ... VIRTUAL TABLE` before the fix.
- [x] `pytest -q tests/test_retro_self_pollution_quarantine.py`
- [x] `ruff check scripts/retro_quarantine_self_pollution.py tests/test_retro_self_pollution_quarantine.py`
- [x] `ruff format --check scripts/retro_quarantine_self_pollution.py tests/test_retro_self_pollution_quarantine.py`
- [x] `pytest -q tests/test_retro_self_pollution_quarantine.py tests/test_search_trigram_fts.py`
- [x] `pytest -q --ignore=tests/test_vector_store.py --ignore=tests/test_engine.py` (`3446 passed, 49 skipped, 5 xfailed`)
- [x] `BRAINLAYER_PREPUSH_SCOPE=changed-only git push -u origin perf/quarantine-rebuild-quadratic-fix` hook (`13 + 3 + 40 pytest tests`, `bun test`, and `test_fts5_determinism.sh` passed)

## Notes
- Did not run the full 17GB rebuild.
- Did not run `--apply` against production.
- Local pre-commit `coderabbit review --agent` was attempted and stopped after the bounded three-minute window with only heartbeat/status output; PR-level bot reviews are requested next.

---

> [!NOTE]
> **High Risk**
> Changes bulk SQLite/FTS5 rewrite logic on the production quarantine path; mistakes could corrupt search indexes or restore metadata on large databases, though tests and revert proofs mitigate this.
> 
> **Overview**
> Reworks **retro self-pollution quarantine apply** so manifest recording and restore snapshots use **`chunk_fts_rowids`** joins instead of **`chunk_id` filters on FTS5 virtual tables**, avoiding quadratic virtual-table scans on large databases.
> 
> **`apply_quarantine_ids`** now syncs the rowid map up front, preloads delete rowids for the full apply set, runs **real per-batch** manifest/reclassify/operational-FTS work with commits and checkpoints, then **rebuilds** knowledge (and trigram) FTS with **manifest rowid reservations** so unquarantine can restore original FTS rowids. **`unquarantine_ids`** and apply both run under **`VectorStore`’s writer pidfile lock** while FTS triggers are dropped.
> 
> **`_record_manifest`** bulk-inserts via a reconcile temp table (`INSERT OR IGNORE` for same-run retries). **`capture_restore_state`** prefers the rowid map with fallbacks for corrupt maps, orphans, and duplicates. Optional trigram tables are detected with **`_table_exists`**.
> 
> Tests add **EXPLAIN QUERY PLAN** guards, incremental-vs-rebuild parity, writer-lock ordering, legacy schemas without trigram, rowid reservation edge cases, and duplicate-FTS cleanup.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 52b04920e31d5b250593ccd38e669d7fbab7009f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Fix quadratic manifest scan in quarantine rebuild by using indexed rowid map joins
> - Replaces full virtual table scans in `_record_manifest` and FTS operations with indexed lookups via `chunk_fts_rowids`, eliminating O(n²) behavior during quarantine rebuilds.
> - Adds `_delete_fts_rows_for_reconcile` to drive targeted FTS deletions from precomputed rowid staging tables, keeping `chunk_fts_rowids` consistent after each batch.
> - `apply_quarantine_ids` and `unquarantine_ids` now run under `_operation_writer_lock` to serialize writes, tolerate absence of `chunks_fts_trigram`, and use `INSERT OR IGNORE` so re-running with the same `run_id` does not overwrite existing manifest rows.
> - Adds `_rebuild_chunk_fts_rowids` to atomically repopulate the rowid map from all present FTS tables after each batch completes.
> - Risk: `_record_manifest` now raises `ValueError` if any chunk_id in the reconcile table is missing from `chunks`, which is a new failure mode compared to prior behavior.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized 52b0492.</sup>
> 
>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved quarantine and restore reliability for databases with missing or varying FTS data.
  * Fixed restore behavior for tables with duplicate or missing search rows, helping preserve the correct search results.
  * Made apply/unquarantine flows more resilient when optional search indexes are unavailable.
  * Reduced the chance of data mismatches during rebuilds and restores after quarantine operations.

### CODE REVIEW

File: scripts/retro_quarantine_self_pollution.py [hunk 1]
```diff
@@ -333,44 +362,145 @@ def capture_restore_state(cursor: apsw.Cursor, chunk_ids: list[str]) -> dict[str
...
+    return [
+        tuple(row)
+        for row in cursor.execute(
+            f"SELECT {FTS_STATE_COLUMNS} FROM {table_name} WHERE chunk_id = ? ORDER BY rowid",
+            (chunk_id,),
+        )
+    ]
+
+
+def _record_manifest(
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py:448]: Fixed in 7e4dfd0e. `_record_manifest` now receives `include_trigram` and emits `NULL` without joining `chunks_fts_trigram` when that optional table is absent. Added `test_apply_without_trigram_table_records_manifest_without_trigram_rowid` to cover `bootstrap_schema=False` legacy DBs with no trigram table.

File: scripts/retro_quarantine_self_pollution.py [hunk 2]
```diff
@@ -424,30 +555,98 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
+        """
+    )
+    cursor.execute(
+        f"""
+        INSERT INTO chunks_fts({FTS_COLUMNS})
+        SELECT {FTS_SELECT_COLUMNS_QUALIFIED}
+        FROM chunks c
+        LEFT JOIN chunk_fts_rowids r ON r.chunk_id = c.id
+        WHERE {NON_OPERATIONAL_FTS_CLASS_SQL}
+          AND r.fts_rowid IS NULL
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in dc6d4b70. The default/trigram rebuild now temporarily reserves active manifest rowids before inserting mapped or repaired rows, treats rows whose stale map conflicts with reserved rowids as needing fresh allocation, then removes only the reservation placeholders. Added `test_rebuild_reserves_manifest_rowids_when_repairing_missing_fts_rows`, which reproduced the rollback constraint failure before the fix.

File: scripts/retro_quarantine_self_pollution.py
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py:1]: Fixed in 042eefc7. `capture_restore_state` now falls back to direct `chunk_id` capture when `chunk_fts_rowids` is missing, null, or stale, so orphaned FTS rows are included in restore snapshots. Added `test_capture_restore_state_includes_orphan_fts_rows_without_rowid_map`.

File: scripts/retro_quarantine_self_pollution.py [hunk 3]
```diff
@@ -424,30 +569,98 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
+        """
+    )
+
+
+def _rebuild_chunk_fts_rowids(cursor: apsw.Cursor, *, include_trigram: bool) -> None:
+    cursor.execute("DELETE FROM chunk_fts_rowids")
+    cursor.execute("""
+        INSERT INTO chunk_fts_rowids(chunk_id, fts_rowid)
+        SELECT chunk_id, rowid FROM chunks_fts WHERE chunk_id IS NOT NULL
+    """)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 042eefc7. `_rebuild_chunk_fts_rowids` now uses `INSERT OR IGNORE` ordered by FTS rowid for default FTS rows, matching the duplicate-tolerant behavior of the other rowid-map rebuild inserts. Added `test_rebuild_chunk_fts_rowids_tolerates_duplicate_default_fts_rows`.

File: scripts/retro_quarantine_self_pollution.py [hunk 4]
```diff
@@ -495,31 +708,41 @@ def apply_quarantine_ids(
     _drop_fts_triggers(cursor)
     _ensure_manifest(cursor)
     timestamp = _utc_now()
+    include_trigram = _table_exists(cursor, "chunks_fts_trigram")
+    table_names = ["chunks_fts", "chunks_fts_operational"]
+    if include_trigram:
+        table_names.append("chunks_fts_trigram")
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 042eefc7. `unquarantine_ids` now detects optional trigram table/rowid-column availability, skips trigram deletes/inserts when absent, and dynamically upserts only existing rowid-map columns. Extended `test_apply_without_trigram_table_records_manifest_without_trigram_rowid` to cover revert.

File: scripts/retro_quarantine_self_pollution.py
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py:1]: Fixed in 042eefc7. `_delete_fts_rows_for_reconcile` still uses rowid-map deletes for the hot path, then does a targeted cleanup for reconciled `chunks_fts_operational` chunk_ids before inserting the single replacement row. Added `test_apply_removes_duplicate_target_operational_fts_rows`.

File: scripts/retro_quarantine_self_pollution.py [hunk 5]
```diff
@@ -424,30 +570,198 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
+        manifest_rowid_column="original_trigram_rowid",
+        run_id=run_id,
+    )
+
+
+def _rebuild_chunk_fts_rowids(cursor: apsw.Cursor, *, include_trigram: bool) -> None:
+    cursor.execute("DELETE FROM chunk_fts_rowids")
+    cursor.execute("""
+        INSERT INTO chunk_fts_rowids(chunk_id, fts_rowid)
+        SELECT chunk_id, rowid FROM chunks_fts WHERE chunk_id IS NOT NULL
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 042eefc7 by the same duplicate-default rowid-map change: `_rebuild_chunk_fts_rowids` uses ordered `INSERT OR IGNORE`, with coverage in `test_rebuild_chunk_fts_rowids_tolerates_duplicate_default_fts_rows`.

File: scripts/retro_quarantine_self_pollution.py [hunk 6]
```diff
@@ -424,27 +581,228 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
+    cursor.execute(
+        f"""
+        INSERT OR IGNORE INTO {table_name}(rowid, {FTS_COLUMNS})
+        SELECT DISTINCT
+            m.{manifest_rowid_column},
+            '', NULL, NULL, NULL, NULL, NULL,
+            ?
+        FROM {QUARANTINE_MANIFEST_TABLE} m
+        WHERE m.run_id = ?
+          AND m.{manifest_rowid_column} IS NOT NULL
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 0c98744f. The FTS rebuild reservation now protects manifest rowids across all recorded runs, not just the current run, and the rebuild treats any manifest-owned rowid as reserved before repairing missing FTS rows. Added `test_rebuild_reserves_prior_run_manifest_rowids`, which reproduced the old-run rollback constraint failure before the fix.

File: tests/test_retro_self_pollution_quarantine.py [hunk 7]
```diff
@@ -155,6 +312,506 @@ def test_quarantine_unquarantine_round_trip_restores_chunk_and_fts_rows(tmp_path
...
+            WHERE chunk_id = 'preserved-missing-fts'
+            """
+        ).fetchone()
+    finally:
+        conn.close()
+
+    assert restored_fts_rowid == original_fts_rowid
+    assert restored_trigram_rowid == original_trigram_rowid
+    assert repaired_rowids[0] != original_fts_rowid
+    assert repaired_rowids[1] != original_trigram_rowid
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/test_retro_self_pollution_quarantine.py:929]: Fixed in 0c98744f. The reservation repair test now asserts the repaired rowid tuple exists and both default/trigram rowids are non-null before checking they avoided the reserved manifest rowids.

File: scripts/retro_quarantine_self_pollution.py [hunk 8]
```diff
@@ -495,31 +853,41 @@ def apply_quarantine_ids(
...
             batches_since_checkpoint += 1
             if batches_since_checkpoint >= checkpoint_every:
                 _checkpoint(cursor)
                 batches_since_checkpoint = 0
         cursor.execute("BEGIN IMMEDIATE")
-        _clear_trigram_fts(cursor)
+        _rebuild_knowledge_fts(cursor, run_id=run_id)
+        if include_trigram:
+            _rebuild_trigram_fts(cursor, run_id=run_id)
+        _rebuild_chunk_fts_rowids(cursor, include_trigram=include_trigram)
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 0c98744f. `apply_quarantine_ids` and `unquarantine_ids` now hold an operation-wide writer pidfile lock around trigger recreation/drop, batch commits/checkpoints, FTS rebuilds, and final optimization. Added apply/revert lock coverage to assert trigger maintenance runs under that lock.

File: tests/test_retro_self_pollution_quarantine.py [hunk 9]
```diff
@@ -41,6 +43,161 @@ def _insert_chunk(
...
+    return details
+
+
+def _assert_uses_indexed_rowid_map(details: list[str]) -> None:
+    joined = "\n".join(details)
+    assert "chunk_fts_rowids" in joined
+    assert "SEARCH" in joined
+    assert "SCAN chunks_fts VIRTUAL TABLE" not in joined
+    assert "SCAN chunks_fts_trigram VIRTUAL TABLE" not in joined
+    assert "SCAN chunks_fts_operational VIRTUAL TABLE" not in joined
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/test_retro_self_pollution_quarantine.py:191]: Fixed in 0c98744f. `_assert_uses_indexed_rowid_map` now requires `SEARCH` and `chunk_fts_rowids` in the same query-plan detail, while keeping the negative virtual-table scan assertions.

File: scripts/retro_quarantine_self_pollution.py [hunk 10]
```diff
@@ -333,44 +363,170 @@ def capture_restore_state(cursor: apsw.Cursor, chunk_ids: list[str]) -> dict[str
...
+                ON f.rowid = r.fts_rowid
+               AND f.chunk_id = c.id
+            {trigram_join}
+            LEFT JOIN chunks_fts_operational o
+                ON o.rowid = r.operational_rowid
+               AND o.chunk_id = c.id
             """,
-            (run_id, chunk_id, row[0], row[1], row[2], row[3], row[4], timestamp),
+            (run_id, timestamp),
+        )
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 0c98744f. `_record_manifest` now uses `INSERT OR IGNORE` so a retry with the same `run_id` preserves the first captured original class/provenance/rowids. Added `test_retry_same_run_preserves_initial_manifest_originals` for a partial-apply retry.

File: scripts/retro_quarantine_self_pollution.py [hunk 11]
```diff
@@ -333,44 +376,170 @@ def capture_restore_state(cursor: apsw.Cursor, chunk_ids: list[str]) -> dict[str
...
+                ON f.rowid = r.{rowid_column}
+               AND f.chunk_id = r.chunk_id
+            WHERE r.chunk_id = ?
+            ORDER BY f.rowid
             """,
             (chunk_id,),
+        )
+    ]
+    if rows:
+        return rows
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 7d24a972. `capture_restore_state` now unions the indexed mapped row lookup with the actual `chunk_id` FTS rows, so duplicate/corrupt FTS rows remain visible to restore snapshots. Added `test_capture_restore_state_includes_duplicate_fts_rows_with_valid_rowid_map`.

File: scripts/retro_quarantine_self_pollution.py [hunk 12]
```diff
@@ -424,27 +594,212 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
         SELECT chunk_id, rowid
         FROM chunks_fts_operational
-        WHERE chunk_id IN ({placeholders})
+        WHERE chunk_id IN (SELECT chunk_id FROM {RECONCILE_CHUNK_ID_TABLE})
+        ON CONFLICT(chunk_id) DO UPDATE SET operational_rowid = excluded.operational_rowid
+        """
+    )
+
+
+def _reserve_manifest_fts_rowids(
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 7d24a972. The rebuild no longer reserves every historical manifest rowid. It materializes a temp reserved-rowid table before deleting each FTS table and includes only manifest rowids whose original FTS row is currently absent, so restored prior-run rows keep their rowids while outstanding prior-run quarantines remain protected. Added `test_rebuild_does_not_reserve_restored_prior_run_manifest_rowids`.

File: scripts/retro_quarantine_self_pollution.py [hunk 13]
```diff
@@ -424,27 +594,212 @@ def _clear_trigram_fts(cursor: apsw.Cursor) -> None:
...
+        INSERT INTO chunks_fts(rowid, {FTS_COLUMNS})
+        SELECT r.fts_rowid, {FTS_SELECT_COLUMNS_QUALIFIED}
+        FROM chunks c
+        INNER JOIN chunk_fts_rowids r ON r.chunk_id = c.id
+        WHERE {NON_OPERATIONAL_FTS_CLASS_SQL}
+          AND r.fts_rowid IS NOT NULL
+          AND NOT EXISTS (
+              SELECT 1
+              FROM {QUARANTINE_MANIFEST_TABLE} m
+              WHERE m.original_fts_rowid = r.fts_rowid
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 7d24a972. The rebuild now materializes outstanding manifest rowids into temp `INTEGER PRIMARY KEY` tables and probes those tables during the rowid-preserving inserts, avoiding correlated scans of the manifest. Added `test_rebuild_fts_uses_indexed_reserved_rowid_table_for_manifest_probes`.

File: scripts/retro_quarantine_self_pollution.py [hunk 14]
```diff
@@ -333,44 +380,173 @@ def capture_restore_state(cursor: apsw.Cursor, chunk_ids: list[str]) -> dict[str
...
+        cursor.execute(
+            f"""
+            DELETE FROM {table_name}
+            WHERE rowid IN (
+                SELECT r.{rowid_column}
+                FROM chunk_fts_rowids r
+                INNER JOIN {RECONCILE_CHUNK_ID_TABLE} q ON q.chunk_id = r.chunk_id
+                WHERE r.{rowid_column} IS NOT NULL
+            )
+            AND chunk_id IN (SELECT chunk_id FROM {RECONCILE_CHUNK_ID_TABLE})
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/retro_quarantine_self_pollution.py]: Fixed in 52b04920. Apply now stages all target FTS rowids once before batching, including duplicate/orphan rows in chunks_fts, chunks_fts_trigram, and chunks_fts_operational, then each committed batch deletes by those staged rowids before marking chunks operational. Added a regression that forces failure before the final rebuild and verifies default/trigram rows are already gone while the operational row remains.
